annotations/src/main/java/com/squareup/anvil/annotations/ContributesMultibinding.kt

```diff
@@ -0,0 +1,90 @@
+
+import kotlin.annotation.AnnotationRetention.RUNTIME
+import kotlin.annotation.AnnotationTarget.CLASS
+import kotlin.reflect.KClass
+
+/**
+ * Generate a Dagger multibinding method for an annotated class and contributes this multibinding
+ * method to the given [scope]. Imagine this example:
+ * ```
+ * interface Listener
+ *
+ * class MainListener @Inject constructor() : Listener
+ *
+ * @Module
+ * @ContributesTo(AppScope::class)
+ * abstract class MainListenerModule {
+ *   @Binds @IntoSet
+ *   abstract fun bindMainListener(listener: MainListener): Listener
+ * }
+ * ```
+ * This is a lot of boilerplate. You can replace this entire module with the
+ * [ContributesMultibinding] annotation. The equivalent would be:
+ * ```
+ * interface Listener
+ *
+ * @ContributesMultibinding(AppScope::class)
+ * class MainListener @Inject constructor() : Listener
+ * ```
+ * Notice that it's optional to specify [boundType], if there is only exactly one super type. If
+ * there are multiple super types, then it's required to specify the parameter:
+ * ```
+ * @ContributesMultibinding(
+ *   scope = AppScope::class,
+ *   boundType = Listener::class
+ * )
+ * class MainListener @Inject constructor() : Activity(), Listener
+ * ```
+ *
+ * [ContributesMultibinding] supports qualifiers. If you annotate the class additionally with a
+ * qualifier, then the generated multibinding method will be annotated with the same qualifier,
+ * e.g.
+ * ```
+ * @ContributesMultibinding(AppScope::class)
+ * @Named("Prod")
+ * class MainListener @Inject constructor() : Listener
+ *
+ * // Will generate this binding method.
+ * @Binds @IntoSet @Named("Prod")
+ * abstract fun bindMainListener(listener: MainListener): Listener
+ * ```
+ *
+ * Contributed multibindings can replace other contributed modules and contributed multibindings
+ * with the [replaces] parameter. This is especially helpful for different multibindings in
+ * tests.
+ * ```
+ * @ContributesMultibinding(
+ *   scope = AppScope::class,
+ *   replaces = [MainListener::class]
+ * )
+ * class FakeListener @Inject constructor() : Listener
+ * ```
+ * [ContributesMultibinding] supports Kotlin objects, e.g.
+ * ```
+ * @ContributesMultibinding(AppScope::class)
+ * object MainListener : Listener
+ * ```
+ * In this scenario instead of generating a `@Binds @IntoSet` method Anvil will generate a
+ * `@Provides @IntoSet` method returning [boundType].
+ *
+ * [ContributesMultibinding] can be used in conjunction with [ContributesBinding]. Each annotation
+ * will generate the respective binding or multibindings method. If a qualifier is only meant for
+ * one of the annotations, then you can set [ignoreQualifier] to `true` and the qualifier won't
+ * be added to the generated multibinding method.
+ * ```
+ * @ContributesBinding(AppScope::class)
+ * @ContributesMultibinding(AppScope::class, ignoreQualifier = true)
+ * @Named("Prod")
+ * object MainListener : Listener
+ * object MainListener : Listener
+ * ```
+ */
```

Comment:
I really enjoyed the API surface, and I like how simple it is to use `Set` or `Map`. I think this feature will help to remove a lot of boilerplate and it is very intuitive in my opinion. Thank you a lot for this work!

There is one concern that I would like to raise around the overall API, when considering #219, that might affect developers that do not have a lot of experience with Dagger: I feel many of them struggle with multi-binding. If I look at vanilla Dagger, it requires a specific annotation (`@IntoMap`, `@IntoSet`) that help developers to instantly identify what type of multi-binding they are using, and it also helps to fail when a wrong combination is found (e.g., `@IntoMap` without a key).

On the other hand, it might be hard for those developers to understand if a `@ContributesMultibinding` is a `Map` or `Set`. Another point is that, correct me if I'm wrong, removing a `@MapKey` from a `@ContributesMultibinding` by mistake would automatically send that contributes to a `Set` without the developer be aware of and could be trick to catch or even result in runtime exceptions when the key is not found in the map.

I personally prefer how you did with `@ContributesMultibinding` and it feels more intuitive to me but I think it is relevant to consider the cases I mentioned above. What do you think? 🤔

Reply:
I thought about that, but I disliked the idea of having two different annotations or adding a parameter to differentiate between a set and map. This parameter felt like boilerplate when it can be implied. The parameter shouldn't use a default value, so you always must specify it. If you set the parameter to a Map and the map key is missing, then it's an error. If you set the parameter to a Set and add a map key, then it's an error. It feels like you need to declare twice whether you want a set or map multibinding. Since Anvil is all about less boilerplate I decided for this way. 

I agree that it can lead to mistakes, but you could say the same thing about multibindings in general, e.g. when you forget to include the module with vanilla Dagger.